custom_components/tahoma/climate.py

```diff
@@ -1,498 +1,34 @@
-    PRESET_AWAY: STATE_PRESET_AWAY,
-    PRESET_FREEZE: STATE_PRESET_FREEZE,
-    PRESET_NONE: STATE_PRESET_MANUAL,
-    PRESET_SLEEP: STATE_PRESET_SLEEPING_MODE,
-}
-AEH_MAP_PRESET_REVERSE = {
-    PRESET_NONE: PRESET_OFF,
-    PRESET_FREEZE: PRESET_FROST_PROTECTION,
-    PRESET_ECO: PRESET_ECO,
-    PRESET_COMFORT: PRESET_COMFORT,
-}
-TAHOMA_TYPE_HEATING_SYSTEM = "HeatingSystem"
+ST = "SomfyThermostat"
+SUPPORTED_CLIMATE_DEVICES = [AEH, ST]
 
 
 async def async_setup_entry(hass, entry, async_add_entities):
-    """Set up the Tahoma sensors from a config entry."""
+    """Set up the TaHoma climate from a config entry."""
 
     data = hass.data[DOMAIN][entry.entry_id]
-
-    entry.add_update_listener(update_listener)
-
-    entities = []
     controller = data.get("controller")
 
-    for device in data.get("devices"):
-        if (
-            TAHOMA_TYPES[device.uiclass] == "climate"
-            and device.widget in SUPPORTED_CLIMATE_DEVICES
-        ):
-            options = dict(entry.options)
-            preset_temp = {
-                PRESET_FREEZE: None,
-                PRESET_AWAY: None,
-                PRESET_ECO: None,
-                PRESET_COMFORT: None,
-            }
-            if DEVICE_CLASS_TEMPERATURE in options:
-                if device.url in options[DEVICE_CLASS_TEMPERATURE]:
-                    sensor_id = options[DEVICE_CLASS_TEMPERATURE][device.url]
-                    if device.widget == ST:
-                        entities.append(
-                            TahomaClimate(device, controller, sensor_id, preset_temp)
-                        )
-                    else:
-                        for k, v in preset_temp.items():
-                            if (
-                                k in options[DEVICE_CLASS_TEMPERATURE]
-                                and options[DEVICE_CLASS_TEMPERATURE][k] != ""
-                            ):
-                                preset_temp[k] = float(
-                                    options[DEVICE_CLASS_TEMPERATURE][k]
-                                )
-                        entities.append(
-                            TahomaClimate(device, controller, sensor_id, preset_temp)
-                        )
-
-            else:
-                entities.append(
-                    TahomaClimate(device, controller, preset_temp=preset_temp)
-                )
+    climate_devices = [
+        d for d in data.get("devices") if TAHOMA_TYPES[d.uiclass] == "climate"
+    ]
 
+    entities = []
+    for device in climate_devices:
+        if device.widget == AEH:
+            entities.append(AtlanticElectricalHeater(device, controller))
+        elif device.widget == ST:
+            base_url = device.url.split("#", 1)[0]
+            sensor = None
+            for k, v in hass.data["entity_registry"].entities.items():
+                if v.unique_id == base_url + "#2":
+                    sensor = k
+                    break
```

Comment:
It seems you can retrieve an entity directly using its id:
```
entity_registry = await self.hass.helpers.entity_registry.async_get_registry()
entity_entry = entity_registry.async_get(f"{base_url}#2")
```

Reply:
done